### DISCUSSION TITLE
x-ray tracing for Prisma postgres

### DISCUSSION THREAD

**Discussion by ls-jingbo-jin** [prisma/prisma#7824] [Q&A]
Hey guys, did anyone try enable x-ray tracing for Prisma ? 
sequelize can do so by [below](https://github.com/aws/aws-xray-sdk-node/issues/271#issuecomment-608013220):
```ts
const sequelize = new Sequelize(config.database_url);

if (xray) {
  Sequelize.useCLS(xray.getNamespace());

  try {
    xray.capturePostgres(sequelize.connectionManager.lib);
  } catch (err) {
    log.error(err);
  }
}
```

And TypeORM can do so by below, since it is using `pg` under the hood:
```
 AWSXRay.capturePostgres(require('pg'));
```

Does Prisma have similar exposed connetionManager?

**Comment by ryands17** [comment]
@ls-jingbo-jin 👋 

Currently not, but you would need to add segments and sub-segments manually when queries are logged [here](https://www.prisma.io/docs/concepts/components/prisma-client/working-with-prismaclient/logging#event-based-logging). I haven't tried it personally but it's something that can be done 🙂

**Reply by ls-jingbo-jin** [reply]
Thanks @ryands17 👍. I will give it a try.

**Reply by ryands17** [reply]
Let me know in the comments how it goes :)

**Reply by kevin-abiera** [reply]
@ryands17, we ended up with something like this:

```ts
      if (xrayIsEnabled()) {
        const segment = AWSXRay.getSegment();
        if (segment) {
          // NOTE: The following implementation needs better support from the Prisma client
          // Better if we can combine the next two in a single subsegment
          // https://github.com/prisma/prisma/discussions/7824
          // https://github.com/prisma/prisma/issues/7596
          // https://github.com/prisma/prisma/issues/6882
          prisma.$use(async (params, next) => {
            const subsegment = segment.addNewSubsegment(
              `prisma query - ${params.model ? params.model + '.' : ''}${params.action}`,
            );

            const result = await next(params);
            subsegment.addAnnotation('prismaVersion', Prisma.prismaVersion.client);
            subsegment.addSqlData({
              action: params.action,
              model: params.model,
            });
            subsegment.close();
            return result;
          });

          prisma.$on('query', (e) => {
            const subsegment = segment.addNewSubsegment('prisma query');

            // Time elapsed (in milliseconds) between client issuing query and database responding - not only time taken to run query
            // See https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#event-types
            subsegment.start_time = (Date.now() - e.duration) / 1000;

            subsegment.addSqlData({
              sanitized_query: e.query,
            });
            subsegment.close();
          });
        }
      }
```

Had to use both `prisma.$use` and `prisma.$on('query', ...` to get the details of the query. Better if we could have them in a single subsegment. Or get them in a single place (https://github.com/prisma/prisma/issues/7596)

**Reply by ryands17** [reply]
Thanks for the approach 🙂

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
